code/modules/power/lighting.dm

```diff
@@ -461,31 +510,33 @@
 				else
 					playsound(loc, 'sound/effects/glasshit.ogg', 90, TRUE)
 		if(BURN)
-			playsound(src.loc, 'sound/items/welder.ogg', 100, TRUE)
+			playsound(loc, 'sound/items/welder.ogg', 100, TRUE)
 
 // returns whether this light has power
 // true if area has power and lightswitch is on
 /obj/machinery/light/proc/has_power()
 	var/area/A = get_area(src)
 	return A.lightswitch && A.power_light
 
-/obj/machinery/light/proc/flicker(var/amount = rand(10, 20))
-	if(flickering) return
-	flickering = 1
-	spawn(0)
-		if(on && status == LIGHT_OK)
-			for(var/i = 0; i < amount; i++)
-				if(status != LIGHT_OK) break
-				on = !on
-				update(0)
-				sleep(rand(5, 15))
-			on = (status == LIGHT_OK)
-			update(0)
-		flickering = 0
+/obj/machinery/light/proc/flicker(amount = rand(10, 20))
+	if(flickering)
+		return
+	if(on)
+		flickering = TRUE
+		for(var/i in 1 to amount)
+			if(status != LIGHT_OK)
+				break
+			on = !on
+			update() // No param means that ghosts can burn out lights
+			sleep(rand(5, 15))
```

Comment:
Please ensure this won't hang the whole code path above.
Revenant spells will also call this for instance

Reply:
Tested both of them with `sleep(200)`, and it only freezes the proc it's called in. You can still call others while it's sleeping, or even call the same proc again.